### ISSUE TITLE
AWS policy allow to receive credentials through configuration

### ISSUE THREAD

**Issue by juanmolle** [envoyproxy/envoy#31403]
Title: AWS policy allow to receive credentials through configuration

Description:
Nowaday AWS_Lambda and AWS_Request_Signing allows to provide credentials from 3 diferent sources
- ENV VAR
- File
- AWS Role
- 
The request is the possibility to add a 4th mechanism to honor credentials from the filter configuration if they are provided.

**Comment by juanmolle** [comment]
@lavignes @suniltheta Do you think this is possible to be included?

**Comment by suniltheta** [comment]
It is possible to add the option to securely send the credentials via xDS configuration. This was probably not a requirement before because of the complexity involved in managing the credentials on the control plane.

Meanwhile distributing the credentials via control plane may not be the best practice given it is better if each entity fetch the particular creds on its own, which conform to how aws cli/sdk fetch them. However, we can also argue that with Envoy the preferred way to distribute anything is via configurations.

Recently (with v1.29) there was an addition of fetching the credentials via AWS STS service (https://github.com/envoyproxy/envoy/pull/31004), which seems like secure way to fetch the necessary credentials from a managed AWS service using a token file to identify itself for now. Is it possible to extend this functionality to provide a distinct token for each Envoy via configuration and they fetch the credentials from STS using that token?

**Comment by juanmolle** [comment]
@suniltheta I'm reviewing this topic. 
Now a day, we need to add filter for different accounts, more than 1. all providers only support 1 set of credentials. I understand the security implication to add credential in the config but could the config contains a profile, instead of using en ENV VAR to select one from a file, use a config to allows each filter instance choose the credentials it want to use.

If there is already a mechanism to use multiple credentials please let me know

**Comment by juanmolle** [comment]
\nostale

**Comment by suniltheta** [comment]
I think it should be possible to add optional credentials profile in configuration and use those profile with help of Credentials File Credentials provider. 
Today we just rely on env AWS_PROFILE which can very well be configured via API.
https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/aws/utility.cc#L467-L478

**Comment by juanmolle** [comment]
Exactly, I see that, but all filter instances will use the same credential. we want to extend to have for example 2 filter with 2 arn the lives in 2 different AWS account with different credentials. each filter could use a different profile from the same config file. and the filter config will not need to contains secrets, the file will rotate, etc.

I don't know how extend this to AWS Roles though

**Comment by suniltheta** [comment]
best way to do this is to add cross account trust policy, where the role assumed by Envoy will contain policies that allow permissions from multiple accounts. https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies-cross-account-resource-access.html

If you still want to put multiple credentials into the aws config & credentials file, then instead of Envoy choosing `[default]` or any other profile specified in env variable AWS_PROFILE, the profile to pick in Credentials File Based Credentials Provider can be sent via filter configuration, for which code change has to be done.

**Comment by nbaws** [comment]
Ill start a PR on specifying profile name explicitly within filter config

**Comment by suniltheta** [comment]
^ https://github.com/envoyproxy/envoy/pull/33003/files is trying to do that

**Comment by nbaws** [comment]
i see. given that it is a per filter configuration we can implement on aws request signer once this is merged.

**Comment by nbaws** [comment]
regarding adding credentials to xDS config - there are two reasons I wouldn't be inclined to implement this:

1. AWS best practices ( https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#bp-workloads-use-roles ) are to use IAM Roles wherever possible. This is because they are effectively dynamic, and have a fixed expiration time after which they are no longer usable. Even though the current model of reading from credential files doesn't exclude using fixed access keys/secret keys, I'd rather not encourage further use of static credentials rather favour the native role based options available.

2. This now creates another location that static credentials are propagated and must be secured. Again, even though the credential provider currently supports reading static credentials from credential files, it's a known quantity and a somewhat legacy scenario within the AWS SDK. I'd rather not encourage another location that static credentials (of which use is discouraged) must be secured.

If you are able to share more about the use case, I may have some other suggestions on how to address.

**Comment by juanmolle** [comment]
The scenario is the following:
Devs running envoy in a dev account want to access prod lambdas, but prod lambdas during dev and test. they don't want to ask infrastructure department each time they create a vm, to assign a role, nor create or modify access roles in prod to let the other account vm (that is created and deleted by demand). then, they can use temporary credentials for the tests.
Devs can config envoy remotely but not the credentials files nor roles. Similar to istio, resources can be created to config the proxy but not access to the pod to change a file.

If we allow to add credentials in the config they can call the lambda without infrastructure asks.

**Comment by nbaws** [comment]
A solution is probably greater than scope of this change. Can you drop me a mail to nbaws@amazon.com and I can discuss with you further some options.
